### PR TITLE
Convert the MCX functions to gates

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

The multi-controlled Toffoli gates are currently implemented as functions only without a underlying gate class. This PR refactors the different implementations to `ControlledGate`s with a `MCXGate` base class.

### Details and comments

The 3- and 4-controlled X gates are implemented as `CCCXGate` and `CCCCXGate` explicitly. 

For the generic multi-controlled X gate we have three (and a half) existing implementations:
* using the Gray code, no ancillas required 
* using recursive splitting until the 3- or 4-controlled X gate is reached, 1 ancilla required
* using a V-chain of Toffolis, using `num_ctrl_qubits - 2` clean ancillas -- here a version for dirty ancillas exist that does some pre- and post-operations on the ancillas

Having circuit synthesis in mind and that the transpiler can choose the implementation based on the availability on ancillas, the gates have been split in four classes,
* a `MCXGate` base class which serves as abstract building block (defaults to the gray code because that does not require ancillas)
* the `MCXGrayCode` 
* the `MCXRecursive`
* the `MCXVChain` (takes a `dirty_ancillas` keyword argument in the initializer)

If a multi-controlled X gate is instantiated with 4 or less controls, where we know the actual definition, the concrete gate is returned. E.g.
```
type(MCXGate(num_ctrl_qubits=2)) == CCXGate  # True
```
### Questions

- [x] Should the 3- and 4-controlled X gates be named C3XGate and C4XGate for readability? 
- [ ] Is the implementation of `MCXGate.__new__` correct? (Probably is, all tests work fine, but just to be sure!)
- [ ] Should `mct` be kept as circuit method or is that too suggestive to be a multi-controlled T gate?

### CODE REVIEW

File: qiskit/extensions/standard/x.py [hunk 1]
```diff
@@ -12,9 +12,9 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 
-"""
-X, CX and CCX gates.
-"""
+"""X, CX and CCX gates."""
```
Comment (ajavadia) [inline qiskit/extensions/standard/x.py:15]:
```suggestion
"""X, CX, CCX, CCCX and CCCCX gates."""
```

File: qiskit/extensions/standard/x.py [hunk 2]
```diff
@@ -414,3 +447,455 @@ def ccx(self, control_qubit1, control_qubit2, target_qubit,
...
+            h d; cu1(-pi/8) b,d; h d;
+            cx b,c;
+            h d; cu1(pi/8) c,d; h d;
+            cx a,c;
+            h d; cu1(-pi/8) c,d; h d;
+            cx b,c;
+            h d; cu1(pi/8) c,d; h d;
+            cx a,c;
+            h d; cu1(-pi/8) c,d; h d;
+        }
```
Comment (ajavadia) [inline qiskit/extensions/standard/x.py:506]: this block seems like it belongs to CCCCX further down

Comment (Cryoris) [reply qiskit/extensions/standard/x.py:506]: Yes, I'll move it there

**Comment by chriseclectic** [comment]
"mcx" (and "mcy", "mcz", "mcu1", "mcu3", "mcswap")  are supported basis gate for Aer, so if these can also be left as those instructions in qobj rather than unrolled it will greatly increase simulation performance.

**Comment by Cryoris** [comment]
Yes, good point I'll change CCCX and CCCCX to be named "mcx". Question: should we add the `.cccx` and `.ccccx` methods? They are currently not used and we could just use `.mcx` instead. That might look less random and more sorted on the circuit.

**Comment by ajavadia** [comment]
is there any scenario where `MCX(3)` or `MCX(4)` get synthesized as the current `CCCX` and `CCCCX`.

**Comment by Cryoris** [comment]
So  the `CCCX` gate is more efficient than using a V-chain or Gray code by at least a factor 2 in the `cx` counts. Once we have 4 controls, the V-chain implementation (with clean ancillas) needs 2 less `cx` gates (18 instead of 20) and could be more efficient. 
So `MCX(3)` should always get synthesized to `CCCX` and `MCX(4)` always to `CCCCX` -- except the V-chain with clean ancillas is chosen.

File: qiskit/extensions/standard/equivalence_library.py [hunk 3]
```diff
@@ -26,7 +26,7 @@
     MSGate,
     RGate,
     RCCXGate,
-    RCCCXGate,
+    RC3XGate,
```
Comment (ajavadia) [inline qiskit/extensions/standard/equivalence_library.py:29]: should add C3X and C4X to the equivalence library as well?

**Comment by Cryoris** [comment]
>is it possible to move MCX out of the gates and to the circuit library?

Yes, I'll move it there. It makes sense to distinguish between the standard gates are the more complex blocks like the MCX. We'll need to keep the `mct` method though and deprecate it.

**Pull request merged**